crates/oxc_cli/src/plugin.rs

```diff
@@ -0,0 +1,355 @@
+                        }
+                    }
+                }
+            };
+        Ok(())
+    }
+
+    pub fn run_tests(
+        &self,
+        ctx: &mut LintContext,
+        relative_file_path_parts: Vec<Option<String>>,
+        rules_to_run: RulesToRun,
+    ) -> oxc_diagnostics::Result<()> {
+        let inner = Adapter::new(Rc::clone(ctx.semantic()), relative_file_path_parts);
+        let adapter = Arc::from(&inner);
+        if let RulesToRun::Only(this_rule) = rules_to_run {
+            for rule in self.rules.iter().filter(|x| x.name == this_rule) {
+                self.run_plugin_rules(ctx, rule, &adapter)?;
+            }
+        } else {
+            for rule in &self.rules {
+                self.run_plugin_rules(ctx, rule, &adapter)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn run_test(
+    test: &SingleTest,
+    rule_name: &str,
+    plugin: &LinterPlugin,
+) -> std::result::Result<Vec<Report>, Vec<Report>> {
+    let file_path = &test.relative_path.last().expect("there to be atleast 1 path part");
+    let source_text = &test.code;
+
+    let allocator = Allocator::default();
+    let source_type = SourceType::from_path(file_path).unwrap();
+    let ret = Parser::new(&allocator, source_text, source_type).parse();
+
+    // Handle parser errors
+
+    if !ret.errors.is_empty() {
+        return Err(ret.errors);
+    }
+
+    let program = allocator.alloc(ret.program);
+    let SemanticBuilderReturn { semantic, errors } =
+        SemanticBuilder::new(source_text, source_type).with_trivias(ret.trivias).build(program);
+
+    if !errors.is_empty() {
+        return Err(errors);
+    }
+
+    let semantic = Rc::new(semantic);
+
+    let mut lint_ctx = LintContext::new(&Rc::clone(&semantic));
+
+    let result = plugin.run_tests(
+        &mut lint_ctx,
+        test.relative_path.iter().map(|el| Some(el.clone())).collect::<Vec<_>>(),
+        RulesToRun::Only(rule_name.to_string()),
+    );
+
+    if let Some(err) = result.err() {
+        return Err(vec![err]);
+    }
+
+    Ok(lint_ctx.into_message().into_iter().map(|m| m.error).collect::<Vec<_>>())
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Test expected to pass, but failed.")]
+struct ExpectedTestToPassButFailed {
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test failed."]
+    err_span: SourceSpan,
+    #[related]
+    errors: Vec<Report>,
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Test expected to fail, but passed.")]
+struct ExpectedTestToFailButPassed {
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test should have failed but it passed."]
+    err_span: SourceSpan,
+}
+
+fn span_of_test_n(query_text: &str, test_ix: usize) -> SourceSpan {
+    let yaml = YamlLoader::load_from_str(
+        // TODO: Should we just save the string after we read it originally?
+        query_text,
```

Comment:
Should we just save the string after we read it originally? This a somewhat rare event so not sure if it's worth it to always store every rule's string value, rather than just reading one or a few twice.